### PR TITLE
Making ops c10-full: list of optional tensors

### PR DESCRIPTION
**Pull request by smessmer**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #49613 Remove backward and requires_grad from Autograd backend key
* #49259 Remove `use_c10_dispatcher: full` lines
* #49251 Remove generated_unboxing_wrappers and setManuallyBoxedKernel
* #49220 Remove .impl_UNBOXED() and functionalities associated with it
* #49164 Remove codegen logic to support non-c10-full ops
* #49619 Enforce c10-fullness for all ops
* **#49138 Making ops c10-full: list of optional tensors**

See for details: https://fb.quip.com/QRtJAin66lPN

We need to model optional types explicitly, mostly for schema inference. So we cannot pass a `Tensor?[]` as `ArrayRef<Tensor>`, instead we need to pass it as an optional type. This PR changes it to `torch::List<c10::optional<Tensor>>`. It also makes the ops c10-full that were blocked by this.

#### Script
```py
from torch.utils.benchmark import Timer

counts = Timer(
    stmt="""
        auto t = {{op call to measure}};
    """,
    setup="""
        using namespace torch::indexing;
        auto x = torch::ones({4, 4, 4});
    """,
    language="cpp",
).collect_callgrind(number=1_000)
print(counts)
```
#### Results
|  Op call                                                              |before   |after   |delta  |      |
|------------------------------------------------------------------------|---------|--------|-------|------|
|x[0] = 1                                                                |11566015 |11566015|0      |0.00% |
|x.index({0})                                                            |6807019  |6801019 |-6000  |-0.09%|
|x.index({0, 0})                                                         |13529019 |13557019|28000  |0.21% |
|x.index({0, 0, 0})                                                      |10677004 |10692004|15000  |0.14% |
|x.index({"..."})                                                        |5512015  |5506015 |-6000  |-0.11%|
|x.index({Slice(None, None, None)})                                      |6866016  |6936016 |70000  |1.02% |
|x.index({None})                                                         |8554015  |8548015 |-6000  |-0.07%|
|x.index({false})                                                        |22400000 |22744000|344000 |1.54% |
|x.index({true})                                                         |27624088 |27264393|-359695|-1.30%|
|x.index({"...", 0, true, Slice(1, None, 2), torch::tensor({1, 2})})|123472000|123463306|-8694|-0.01%|

#### Script
```py
from torch.utils.benchmark import Timer

counts = Timer(
    stmt="""
        auto t = {{op call to measure}};
    """,
    setup="""
        using namespace torch::indexing;
        auto x = torch::ones({4, 4, 4}, torch::requires_grad());
    """,
    language="cpp",
).collect_callgrind(number=1_000)
print(counts)
```
Note: the script measures the **forward** path of an op call with autograd enabled (i.e. calls into VariableType). It does not measure the backward path.

#### Results
|  Op call                                                              |before   |after   |delta  |      |
|------------------------------------------------------------------------|---------|--------|-------|------|
|x.index({0})                                                            |14839019|14833019|-6000| 0.00% |
|x.index({0, 0})                                                         |28342019|28370019|28000| 0.00% |
|x.index({0, 0, 0})                                                      |24434004|24449004|15000| 0.00% |
|x.index({"..."})                                                       |12773015|12767015|-6000| 0.00% |
|x.index({Slice(None, None, None)})                                      |14837016|14907016|70000| 0.47% |
|x.index({None})                                                        |15926015|15920015|-6000| 0.00% |
|x.index({false})                                                        |36958000|37477000|519000| 1.40% |
|x.index({true})                                                         |41971408|42426094|454686| 1.08% |
|x.index({"...", 0, true, Slice(1, None, 2), torch::tensor({1, 2})}) |168184392|164545682|-3638710| -2.16% |

## Backwards Compatibility

- This should not break the Python API because the representation in Python is the same and python_arg_parser just transforms the python list into a `List<optional<Tensor>>` instead of into a `List<Tensor>`.
- This should not break serialized models because there's some logic that allows loading a serialized `List<Tensor>` as `List<optional<Tensor>>`, see https://github.com/pytorch/pytorch/pull/49138/files#diff-9315f5dd045f47114c677174dcaa2f982721233eee1aa19068a42ff3ef775315R57
- This will break backwards compatibility for the C++ API. There is no implicit conversion from `ArrayRef<Tensor>` (which was the old argument type) to `List<optional<Tensor>>`. One common call pattern is `tensor.index({indices_tensor})`, where indices_tensor is another `Tensor`, and that will continue working because the `{}` initializer_list constructor for `List<optional<Tensor>>` can take `Tensor` elements that are implicitly converted to `optional<Tensor>`, but another common call pattern was `tensor.index(indices_tensor)`, where previously, the `Tensor` got implicitly converted to an `ArrayRef<Tensor>`, and to implicitly convert `Tensor -> optional<Tensor> -> List<optional<Tensor>>` would be two implicit conversions. C++ doesn't allow chaining. two implicit conversions. So those call sites have to be rewritten to `tensor.index({indices_tensor})`.

Differential Revision: [D25454632](https://our.internmc.facebook.com/intern/diff/D25454632/)

### CODE REVIEW

**Review by bhosmer — APPROVED** [review]
Accepting to unblock, but there's a couple things noted inline that I think should happen before landing - one that might be modestly nontrivial is in the codegen. 

Other than the code, seems worth checking for autograd regressions on optional-list-using ops, if it's not too involved to trigger the codepaths.

Test failures seem like they could be legit - is the XLA similar to the previous thing?

File: aten/src/ATen/native/IndexingUtils.h
Comment (smessmer) [reply aten/src/ATen/native/IndexingUtils.h:52]: I changed it above but not here since this function is very short and would become longer, also there's no repeated `operator->` here.

File: aten/src/ATen/native/LinearAlgebra.cpp
Comment (smessmer) [reply aten/src/ATen/native/LinearAlgebra.cpp:27]: I moved it to IndexingUtils.h since all current use cases are indexing related. Not super happy with it since it is probable that we'll need it for non-indexing-related stuff later on, open for better suggestions.

File: caffe2/contrib/aten/aten_op.cc
Comment (smessmer) [reply caffe2/contrib/aten/aten_op.cc:11]: good catch

File: caffe2/contrib/aten/gen_op.py
Comment (smessmer) [reply caffe2/contrib/aten/gen_op.py:299]: copy&pasted from the comment above, lol.

Seems this is from https://github.com/pytorch/pytorch/pull/8226, not sure what the comment was meant to say originally cc @jamesr66a

File: tools/autograd/templates/Functions.h
Comment (smessmer) [reply tools/autograd/templates/Functions.h:36]: removed

File: tools/codegen/api/python.py
Comment (smessmer) [reply tools/codegen/api/python.py:630]: I don't think this was `TensorList?` once it reached c++. We likely removed the `?`.

File: tools/codegen/api/python.py
Comment (smessmer) [reply tools/codegen/api/python.py:1138]: typo

File: torch/csrc/autograd/VariableTypeManual.cpp
Comment (smessmer) [reply torch/csrc/autograd/VariableTypeManual.cpp:70]: it's not

File: caffe2/contrib/aten/gen_op.py
Comment (jamesr66a) [reply caffe2/contrib/aten/gen_op.py:299]: Yeah, I definitely don't remember what the rest of the comment was meant to say :p

File: tools/autograd/gen_autograd.py
Comment (smessmer) [reply tools/autograd/gen_autograd.py:104]: fixed. it's not special cased anymore.

File: tools/autograd/gen_python_functions.py
Comment (smessmer) [reply tools/autograd/gen_python_functions.py:477]: found a better implementation for it, this is removed.

**Comment by smessmer** [comment]
> Test failures seem like they could be legit - is the XLA similar to the previous thing?

XLA changes are addressed in https://github.com/pytorch/xla/pull/2698

File: caffe2/contrib/aten/gen_op.py
Comment (smessmer) [reply caffe2/contrib/aten/gen_op.py:299]: I removed it ;)

File: tools/codegen/api/cpp.py
Comment (ljk53) [inline tools/codegen/api/cpp.py:109]: Shouldn't this be `ConstRefCType(BaseCType("c10::List<c10::optional<Tensor>>", binds))` instead?

File: tools/codegen/api/python.py
Comment (ljk53) [inline tools/codegen/api/python.py:231]: hmm....

File: tools/codegen/api/cpp.py
Comment (smessmer) [reply tools/codegen/api/cpp.py:109]: @ljk53 fixed this in a separate PR, thanks.

File: tools/codegen/api/python.py
Comment (smessmer) [reply tools/codegen/api/python.py:231]: Agreed. The reason this is required is that the python arg parser tokenizes types with spaces wrongly, it only supports typenames without spaces. So the correct way of fixing this would be to fix the tokenizer in PythonArgParser.

**Pull request closed without merging**
